fix(cli): only publish if changes detected since last release

esolitos opened the pull request:
## What

Changes on the "Publish Vespa CLI" Github workflow:

1. The publish job now checks if there are any changes between the last release tag and the new version being published. This is done by using the `git diff` command to compare the changes in the `client/go` directory.
2. If no changes are detected, the workflow will skip the publish step.
3. The workflow also handles the case where a dev version (starting with "0.0.0-dev") is being used, in which case the changes check is skipped.

## Why

This ensures that the CLI is only published when there are actual changes, preventing unnecessary releases.

A further reason is that we had 2 separate systems checking for changes, one for the Github Releases and another one for Homebrew, this leads to inconsistencies, so this PR merges the validation into an initial check.


## Additional info

- Fixes #32365

esolitos committed:
fix(cli): only publish if changes detected since last release

The changes in this commit focus on the CI workflow for publishing the CLI. The main changes are:

1. The publish job now checks if there are any changes between the last release tag and the new version being published. This is done by using the `git diff` command to compare the changes in the `client/go` directory.
2. If no changes are detected, the workflow will skip the publish step. This ensures that the CLI is only published when there are actual changes, preventing unnecessary releases.
3. The workflow also handles the case where a dev version (starting with "0.0.0-dev") is being used, in which case the changes check is skipped.

freva commented on .github/workflows/publish-cli.yml:
```diff
@@ -80,8 +112,8 @@ jobs:
-    # Publish the CLI when a tag is pushed or a workflow is triggered manually.
-    if: contains(fromJSON('["push", "workflow_dispatch"]'), github.event_name)
+    # Publish the CLI only if changes were detected and the event was a push or workflow_dispatch.
+    if: needs.prepare.outputs.changes-detected == 'true' && contains(fromJSON('["push", "workflow_dispatch"]'), github.event_name)
```
Maybe not require a diff if this is a workflow_dispatch?

esolitos committed:
feat(publish-cli): Checkout repository at tagged version and bypass diff checks

The changes in this commit address the following:

1. Checkout the repository at the tagged version to ensure the correct code is used for the release.
2. Bypass the diff checks when publishing to GitHub and Homebrew, as these checks are already performed in the "prepare" job.

esolitos committed:
feat(ci): publish GitHub release and Homebrew package

This change updates the GitHub Actions workflow to publish the Vespa CLI to both GitHub Releases and Homebrew. The key changes are:

- Publish the GitHub release and capture the release URL
- Only publish to Homebrew if a GitHub release was successful
- Install the CLI package via Homebrew if a release was published

esolitos committed:
fixes after testing the workflow
